interp/interp.go

```diff
@@ -0,0 +1,165 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package interp
+
+import (
+	"fmt"
+	"math"
+	"sort"
+
+	"gonum.org/v1/gonum/spatial/r1"
+)
+
+// Interpolator1D represents a 1D interpolator.
+// It interpolates a y(x) function based on some data over a range [begin(), end()].
+// Both begin() and end() can be infinite, but begin() must be lower than end().
+type Interpolator1D interface {
+
+	// Returns the closed interval over which one can interpolate.
+	Interval() r1.Interval
+
+	// Evaluates interpolated sequence at x in [begin(), end()].
+	// Panics if the argument is outside this range.
+	Eval(float64) float64
+}
+
+// ConstInterpolator1D is a constant 1D interpolator,
+// It is defined over the range [-infinity, infinity].
+type ConstInterpolator1D struct {
+	// Constant Y value.
+	Value float64
+}
+
+// Interval implements Interpolator1D.Interval.
+func (ci ConstInterpolator1D) Interval() r1.Interval {
+	return r1.Interval{Min: math.Inf(-1), Max: math.Inf(1)}
+}
+
+// Eval implements Interpolator1D.Eval.
+func (ci ConstInterpolator1D) Eval(x float64) float64 {
+	return ci.Value
+}
+
+// findSegment returns a tuple of: (i such that xs[i] <= x < xs[i + 1], xs[i]),
+// or panics if such i is not found. Assumes that len(xs) >= 2 without checking.
+func findSegment(xs []float64, x float64) (int, float64) {
+	// Find minimum i s.t. xs[i] >= x, or len(xs) if not found.
+	n := len(xs)
+	i := sort.Search(n, func(i int) bool { return xs[i] > x })
+	if i < n {
+		if i == 0 {
+			// x < begin()
+			panic(fmt.Sprintf("interp: x value %g below lower bound %g", x, xs[0]))
+		} else {
+			return i - 1, xs[i-1]
+		}
+	} else {
+		if xs[n-1] == x {
+			return n - 1, x
+		}
+		panic(fmt.Sprintf("interp: x value %g above upper bound %g", x, xs[n-1]))
+	}
+}
+
+// LinearInterpolator1D is a piecewise linear 1D interpolator, defined
+// over the range [xs[0], xs[len(xs) - 1]].
```

Comment:
```
// Linear1D is a piecewise linear 1-dimensional interpolator.
```

The inclusion of `[xs[0], xs[len(xs) - 1]]` does not help in the docs here since the users will not see that.